Mage.Tests/src/test/java/org/mage/test/cards/cost/modification/CostReduceForEachTest.java

```diff
@@ -283,7 +283,7 @@ public void test_InspiringStatuary_PayByArtifacts() {
 
         checkPlayableAbility("can play", 1, PhaseStep.PRECOMBAT_MAIN, playerA, "Cast Keeper of Tresserhorn", true);
         castSpell(1, PhaseStep.PRECOMBAT_MAIN, playerA, "Keeper of Tresserhorn");
-        addTarget(playerA, "Inspiring Statuary", 5); // as pay
+        addTarget(playerA, "Inspiring Statuary", 4); // (4th one is autochosen as only option left) as pay
```

Comment:
Here is the test with strange manually 4 of 5 and auto 5.

Reply:
The improvise ability is implemented as min=1 max=num_artifacts:
https://github.com/magefree/mage/blob/a673ebffc5ff32d3da411389cdd6a493da35bed2/Mage/src/main/java/mage/abilities/keyword/ImproviseAbility.java#L91

The was designed to not autochoose for effects like this, ones that say "up to".

It can only choose the last one since min=1 and max=1